dev/TEAL.md

```diff
@@ -0,0 +1,250 @@
+| `==` | A is equal to B => {0 or 1} |
+| `!=` | A is not equal to B => {0 or 1} |
+| `!` | X == 0 yields 1; else 0 |
+| `len` | yields length of byte value |
+| `itob` | converts uint64 to big endian bytes |
+| `btoi` | converts bytes as big endian to uint64 |
+| `%` | A modulo B. Panic if B == 0. |
+| `\|` | A bitwise-or B |
+| `&` | A bitwise-and B |
+| `^` | A bitwise-xor B |
+| `~` | bitwise invert value |
+| `mulw` | A times B out to 128-bit long result as low (top) and high uint64 values on the stack |
+
+### Loading Values
+
+Opcodes for getting data onto the stack.
+
+Some of these have immediate data in the byte or bytes after the opcode.
+
+| Op | Description |
+| --- | --- |
+| `intcblock` | load block of uint64 constants |
+| `intc` | push value from uint64 constants to stack by index into constants |
+| `intc_0` | push uint64 constant 0 to stack |
+| `intc_1` | push uint64 constant 1 to stack |
+| `intc_2` | push uint64 constant 2 to stack |
+| `intc_3` | push uint64 constant 3 to stack |
+| `bytecblock` | load block of byte-array constants |
+| `bytec` | push bytes constant to stack by index into constants |
+| `bytec_0` | push bytes constant 0 to stack |
+| `bytec_1` | push bytes constant 1 to stack |
+| `bytec_2` | push bytes constant 2 to stack |
+| `bytec_3` | push bytes constant 3 to stack |
+| `arg` | push LogicSig.Args[N] value to stack by index |
+| `arg_0` | push LogicSig.Args[0] to stack |
+| `arg_1` | push LogicSig.Args[1] to stack |
+| `arg_2` | push LogicSig.Args[2] to stack |
+| `arg_3` | push LogicSig.Args[3] to stack |
+| `txn` | push field from current transaction to stack |
+| `gtxn` | push field to the stack from a transaction in the current transaction group |
+| `global` | push value from globals to stack |
+| `load` | copy a value from scratch space to the stack |
+| `store` | pop a value from the stack and store to scratch space |
+
+**Transaction Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Sender | []byte |
+| 1 | Fee | uint64 |
+| 2 | FirstValid | uint64 |
+| 3 | FirstValidTime | uint64 |
+| 4 | LastValid | uint64 |
+| 5 | Note | []byte |
+| 6 | Lease | []byte |
+| 7 | Receiver | []byte |
+| 8 | Amount | uint64 |
+| 9 | CloseRemainderTo | []byte |
+| 10 | VotePK | []byte |
+| 11 | SelectionPK | []byte |
+| 12 | VoteFirst | uint64 |
+| 13 | VoteLast | uint64 |
+| 14 | VoteKeyDilution | uint64 |
+| 15 | Type | []byte |
+| 16 | TypeEnum | uint64 |
+| 17 | XferAsset | uint64 |
+| 18 | AssetAmount | uint64 |
+| 19 | AssetSender | []byte |
+| 20 | AssetReceiver | []byte |
+| 21 | AssetCloseTo | []byte |
+| 22 | GroupIndex | uint64 |
+| 23 | TxID | []byte |
+
+
+Additional details in the [opcodes document](TEAL_opcodes.md#txn) on the `txn` op.
+
+**Global Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | MinTxnFee | uint64 |
+| 1 | MinBalance | uint64 |
+| 2 | MaxTxnLife | uint64 |
+| 3 | ZeroAddress | []byte |
+| 4 | GroupSize | uint64 |
```

Comment:
Should GroupSize be moved to `Transaction field` as it's specific to the transaction group (similarly to `GroupIndex`)

Alternatively maybe change to "Other fields" (rather than "Global")? In any case, need some additional text to explain what this field is.

Reply:
I feel like `GroupSize` isn't a property of any one `txn` and so I moved it out into 'the environment in which the txn exists' that I called `global`.

Reply:
Ok